File: Week4/VolvoCodingChallenge/app/src/main/java/com/yeah/ruisu/volvocodingchallenge/AppController.java
package com.yeah.ruisu.volvocodingchallenge;

import android.app.Application;

import com.yeah.ruisu.volvocodingchallenge.di.DaggerMainComponent;
import com.yeah.ruisu.volvocodingchallenge.di.MainComponent;
import com.yeah.ruisu.volvocodingchallenge.di.MainModule;


public class AppController extends Application
{
    private MainComponent MyMainComponent;

    public MainComponent getAppComponent()
    {
        if (MyMainComponent == null)
        {
            MyMainComponent = DaggerMainComponent.builder()
                                                .mainModule(new MainModule(this))
                                                .build();
        }
        return MyMainComponent;
    }
}

File: week5/RealVolvoCodingChallenge/app/src/main/java/com/yeah/ruisu/realvolvocodingchallenge/AppController.kt
package com.yeah.ruisu.realvolvocodingchallenge

import android.app.Application
import com.yeah.ruisu.realvolvocodingchallenge.di.application.ApplicationComponent
import com.yeah.ruisu.realvolvocodingchallenge.di.application.ApplicationModule
import com.yeah.ruisu.realvolvocodingchallenge.di.application.DaggerApplicationComponent

class AppController : Application()
{
    private lateinit var applicationComponent: ApplicationComponent

    fun getComponent() : ApplicationComponent
    {
        return if (this::applicationComponent.isInitialized)
        {
            applicationComponent
        }
        else
        {
            createApplicationComponent()
        }
    }

    private fun createApplicationComponent(): ApplicationComponent
    {
        applicationComponent = DaggerApplicationComponent.builder()
                .applicationModule(ApplicationModule(this))
                .build()

        return applicationComponent
    }
}